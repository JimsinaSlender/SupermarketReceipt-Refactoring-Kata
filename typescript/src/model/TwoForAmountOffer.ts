import { Product } from "./Product"
import { Discount } from "./Discount"
import { Offer } from "./Offer"

export class TwoForAmountOffer {
  private minimumQuantityForOffer: number = 2;
  private product: Product;
  private unitPrice: number;

  public constructor(product: Product, unitPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
  }

  public getDiscount(quantityAsInt: number, offer: Offer, quantity: number) {
    const total = offer.argument * Math.floor(quantityAsInt / this.minimumQuantityForOffer) + quantityAsInt % 2 * this.unitPrice;
    const discountAmount = this.unitPrice * quantity - total;
    return new Discount(this.product, "2 for " + offer.argument, discountAmount);
  }
}