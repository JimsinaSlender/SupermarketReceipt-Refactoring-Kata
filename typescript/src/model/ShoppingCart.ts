import { Product } from "./Product"
import { SupermarketCatalog } from "./SupermarketCatalog"
import * as _ from "lodash"
import { ProductQuantity } from "./ProductQuantity"
import { Discount } from "./Discount"
import { Receipt } from "./Receipt"
import { Offer } from "./Offer"
import { SpecialOfferType } from "./SpecialOfferType"
import { TwoForAmountOffer } from "./TwoForAmountOffer";
import { ThreeForTwoOffer } from "./ThreeForTwoOffer";

type ProductQuantities = { [productName: string]: ProductQuantity }
export type OffersByProduct = { [productName: string]: Offer };

export class ShoppingCart {

    private readonly items: ProductQuantity[] = [];
    _productQuantities: ProductQuantities = {};


    getItems(): ProductQuantity[] {
        return _.clone(this.items);
    }

    addItem(product: Product): void {
        this.addItemQuantity(product, 1.0);
    }

    productQuantities(): ProductQuantities {
        return this._productQuantities;
    }

    getQuantityOf(product: Product): ProductQuantity {
        return this._productQuantities[product.name];
    }


    public addItemQuantity(product: Product, quantity: number): void {
        let productQuantity = new ProductQuantity(product, quantity)
        this.items.push(productQuantity);
        let currentQuantity = this._productQuantities[product.name]
        if (currentQuantity) {
            this._productQuantities[product.name] = this.increaseQuantity(product, currentQuantity, quantity);
        } else {
            this._productQuantities[product.name] = productQuantity;
        }

    }

    private increaseQuantity(product: Product, productQuantity: ProductQuantity, quantity: number) {
        return new ProductQuantity(product, productQuantity.quantity + quantity)
    }

    handleOffers(receipt: Receipt, offers: OffersByProduct, catalog: SupermarketCatalog): void {
        for (const productName in this.productQuantities()) {
            const productQuantity = this._productQuantities[productName]
            const product = productQuantity.product;

            if (offers[productName]) {
                const offer: Offer = offers[productName];
                const unitPrice: number = catalog.getUnitPrice(product);

                const offerType = offer.offerType;

                let discount = this.getDiscountForOffer(productName, offerType, offer, unitPrice, product);

                if (discount != null)
                    receipt.addDiscount(discount);
            }

        }
    }

    private getDiscountForOffer(productName: string, offerType: SpecialOfferType, offer: Offer, unitPrice: number, product: Product) {
        const quantity: number = this._productQuantities[productName].quantity;
        const quantityAsInt = quantity;
        let minimumQuantityForOffer = this.getQuantityForOffer(offerType);
        const maybeDiscountMultiple = Math.floor(quantityAsInt / minimumQuantityForOffer);
        let discount: Discount | null = null;

        discount = this.getTwoForAmountDiscount(offerType, quantityAsInt, offer, minimumQuantityForOffer, unitPrice, quantity, discount, product);
        discount = this.threeForTwoDiscount(offerType, quantityAsInt, quantity, unitPrice, maybeDiscountMultiple, discount, product);
        discount = this.tenPercentDiscount(offerType, discount, product, offer, quantity, unitPrice);
        discount = this.fiveForAmountDiscount(offerType, quantityAsInt, unitPrice, quantity, offer, maybeDiscountMultiple, discount, product, minimumQuantityForOffer);
        return discount;
    }

    private fiveForAmountDiscount(offerType: SpecialOfferType, quantityAsInt: number, unitPrice: number, quantity: number, offer: Offer, maybeDiscountMultiple: number, discount: Discount | null, product: Product, minimumQuantityForOffer: number) {
        if (offerType == SpecialOfferType.FiveForAmount && quantityAsInt >= 5) {
            const discountAmount = unitPrice * quantity - (offer.argument * maybeDiscountMultiple + quantityAsInt % 5 * unitPrice);
            discount = new Discount(product, minimumQuantityForOffer + " for " + offer.argument, discountAmount);
        }
        return discount;
    }

    private tenPercentDiscount(offerType: SpecialOfferType, discount: Discount | null, product: Product, offer: Offer, quantity: number, unitPrice: number) {
        if (offerType == SpecialOfferType.TenPercentDiscount) {
            discount = new Discount(product, offer.argument + "% off", quantity * unitPrice * offer.argument / 100.0);
        }
        return discount;
    }

    private threeForTwoDiscount(offerType: SpecialOfferType, quantityAsInt: number, quantity: number, unitPrice: number, maybeDiscountMultiple: number, discount: Discount | null, product: Product) {
        const threeForTwo: ThreeForTwoOffer = new ThreeForTwoOffer(product, unitPrice);

        if (!threeForTwo.applies(this)) {
            return discount;
        }


        return threeForTwo.getDiscount(quantity);
    }

    private getTwoForAmountDiscount(offerType: SpecialOfferType, quantityAsInt: number, offer: Offer, minimumQuantityForOffer: number, unitPrice: number, quantity: number, discount: Discount | null, product: Product) {
        if (!(offerType == SpecialOfferType.TwoForAmount && quantityAsInt >= 2)) {
            return discount;
        }

        const twoForAmount: TwoForAmountOffer = new TwoForAmountOffer(product, unitPrice);
        return twoForAmount.getDiscount(quantityAsInt, offer, quantity);
    }

    private getQuantityForOffer(offerType: SpecialOfferType) {
        let minimumQuantityForOffer = 1;
        if (offerType == SpecialOfferType.ThreeForTwo) {
            minimumQuantityForOffer = 3;
        }
        else if (offerType == SpecialOfferType.TwoForAmount) {
            minimumQuantityForOffer = 2;
        }
        if (offerType == SpecialOfferType.FiveForAmount) {
            minimumQuantityForOffer = 5;
        }
        return minimumQuantityForOffer;
    }
}
