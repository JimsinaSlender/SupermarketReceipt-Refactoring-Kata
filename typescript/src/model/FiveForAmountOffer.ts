import { Product } from "./Product"
import { Discount } from "./Discount"
import { ShoppingCart } from "./ShoppingCart";

export class FiveForAmountOffer {
  private minimumQuantityForOffer: number = 5;
  private product: Product;
  private unitPrice: number;
  private discountedPrice: number;

  public constructor(product: Product, unitPrice: number, discountedPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.discountedPrice = discountedPrice;
  }

  public getDiscount(cart: ShoppingCart) {
    const quantity = cart.getQuantityOf(this.product).quantity;
    const total = this.discountedPrice * Math.floor(quantity / this.minimumQuantityForOffer) + quantity % this.minimumQuantityForOffer * this.unitPrice;
    const discountAmount = this.unitPrice * quantity - total;
    return new Discount(this.product, `${this.minimumQuantityForOffer} for ${this.discountedPrice}`, discountAmount);
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product).quantity >= this.minimumQuantityForOffer;
  }
}