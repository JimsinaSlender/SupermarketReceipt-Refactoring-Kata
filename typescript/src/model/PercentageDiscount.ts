import { ShoppingCart } from "./ShoppingCart";
import { Product } from "./Product";
import { Discount } from "./Discount";

export class PercentageDiscount {
  private product: Product;
  private unitPrice: number;
  private percentageOff: number;

  public constructor(product: Product, unitPrice: number, percentageOff: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.percentageOff = percentageOff;
  }

  public getDiscount(quantity: number): Discount {
    return new Discount(this.product, this.percentageOff + "% off", (quantity * this.unitPrice * this.percentageOff) / 100 );
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product).quantity >= 1;
  }
}