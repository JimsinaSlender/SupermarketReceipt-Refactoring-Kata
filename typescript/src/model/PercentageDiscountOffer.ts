import { ShoppingCart } from "./ShoppingCart";
import { Product } from "./Product";
import { Discount } from "./Discount";

export class PercentageDiscountOffer {
  private product: Product;
  private unitPrice: number;
  private percentageOff: number;

  public constructor(product: Product, unitPrice: number, percentageOff: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.percentageOff = percentageOff;
  }

  public getDiscount(cart: ShoppingCart): Discount {
    const quantity = cart.getQuantityOf(this.product).quantity;
    return new Discount(this.product, this.percentageOff + "% off", (quantity * this.unitPrice * this.percentageOff) / 100 );
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product).quantity >= 1;
  }
}