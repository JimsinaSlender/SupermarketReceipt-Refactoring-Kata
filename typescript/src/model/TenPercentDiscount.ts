import { ShoppingCart } from "./ShoppingCart";
import { Product } from "./Product";
import { Discount } from "./Discount";

export class TenPercentDiscount {
  private product: Product;
  private unitPrice: number;

  public constructor(product: Product, unitPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
  }

  public getDiscount(quantity: number): Discount {
    return new Discount(this.product, "10% off", quantity * this.unitPrice * .1);
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product).quantity >= 1;
  }
}