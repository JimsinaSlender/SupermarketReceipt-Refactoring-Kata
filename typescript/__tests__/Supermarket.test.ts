import { FakeCatalog } from "./FakeCatalog"
import { Product } from "../src/model/Product"
import { SupermarketCatalog } from "../src/model/SupermarketCatalog"
import { Receipt } from "../src/model/Receipt"
import { ShoppingCart } from "../src/model/ShoppingCart"
import { Teller } from "../src/model/Teller"
import { SpecialOfferType } from "../src/model/SpecialOfferType"
import { ProductUnit } from "../src/model/ProductUnit"

describe('Supermarket', () => {
    const applePrice: number = 1.99;
    const toothbrushPrice: number = 0.99;
    const ricePrice: number = 2.49;
    const toothpastePrice: number = 1.79;
    const cherryTomatoesPrice: number = 0.69;
    let toothbrush: Product;
    let toothpaste: Product;
    let apples: Product;
    let rice: Product;
    let cherryTomatoes: Product;
    let catalog: SupermarketCatalog;
    let cart: ShoppingCart;
    let teller: Teller;
    let receipt: Receipt;

    beforeEach(() => {
        catalog = new FakeCatalog();
        toothbrush = new Product("toothbrush", ProductUnit.Each);
        apples = new Product("apples", ProductUnit.Kilo);
        rice = new Product("rice", ProductUnit.Each);
        toothpaste = new Product("toothpaste", ProductUnit.Each);
        cherryTomatoes = new Product("cherry tomatoes", ProductUnit.Each);
        catalog.addProduct(toothbrush, toothbrushPrice);
        catalog.addProduct(apples, applePrice);
        catalog.addProduct(rice, ricePrice);
        catalog.addProduct(toothpaste, toothpastePrice);
        catalog.addProduct(cherryTomatoes, cherryTomatoesPrice);
        cart = new ShoppingCart();
        teller = new Teller(catalog);
    });

    afterEach(() => {
        expect(receipt).toMatchSnapshot();
    });

    it('applies three for two discount', () => {
        cart.addItemQuantity(toothbrush, 3);

        teller.addSpecialOffer(SpecialOfferType.ThreeForTwo, toothbrush, 10.0);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(toothbrushPrice * 2);
    });

    it('applies twenty percent discount', () => {
        const discountPercentage = 20.0;
        const discountMultiplier = (100 - discountPercentage) / 100;
        cart.addItemQuantity(apples, 1);

        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, apples, discountPercentage);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(applePrice * discountMultiplier);
    });

    it('applies ten percent discount', () => {
        const discountPercentage = 10.0;
        const discountMultiplier = (100 - discountPercentage) / 100;
        cart.addItemQuantity(rice, 1);

        teller.addSpecialOffer(SpecialOfferType.TenPercentDiscount, rice, discountPercentage);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(ricePrice * discountMultiplier);
    });

    it('applies five for amount discount', () => {
        const discountedPriceForFive: number = 7.49;
        cart.addItemQuantity(toothpaste, 5);

        teller.addSpecialOffer(SpecialOfferType.FiveForAmount, toothpaste, discountedPriceForFive);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(discountedPriceForFive);
    });

    it('applies two for amount discount', () => {
        const discountedPriceForTwo: number = 0.99;
        cart.addItemQuantity(cherryTomatoes, 2);

        teller.addSpecialOffer(SpecialOfferType.TwoForAmount, cherryTomatoes, discountedPriceForTwo);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(discountedPriceForTwo);
    });
});
