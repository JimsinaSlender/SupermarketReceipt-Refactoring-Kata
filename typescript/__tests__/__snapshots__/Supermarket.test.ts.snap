// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supermarket applies five for amount discount 1`] = `
Receipt {
  "discounts": Array [
    Discount {
      "description": "5 for 7.49",
      "discountAmount": 1.459999999999999,
      "product": Product {
        "name": "toothpaste",
        "unit": 1,
      },
    },
  ],
  "items": Array [
    ReceiptItem {
      "price": 1.79,
      "product": Product {
        "name": "toothpaste",
        "unit": 1,
      },
      "quantity": 5,
      "totalPrice": 8.95,
    },
  ],
}
`;

exports[`Supermarket applies ten percent discount 1`] = `
Receipt {
  "discounts": Array [
    Discount {
      "description": "10% off",
      "discountAmount": 0.24900000000000003,
      "product": Product {
        "name": "rice",
        "unit": 1,
      },
    },
  ],
  "items": Array [
    ReceiptItem {
      "price": 2.49,
      "product": Product {
        "name": "rice",
        "unit": 1,
      },
      "quantity": 1,
      "totalPrice": 2.49,
    },
  ],
}
`;

exports[`Supermarket applies three for two discount 1`] = `
Receipt {
  "discounts": Array [
    Discount {
      "description": "3 for 2",
      "discountAmount": 0.9899999999999998,
      "product": Product {
        "name": "toothbrush",
        "unit": 1,
      },
    },
  ],
  "items": Array [
    ReceiptItem {
      "price": 0.99,
      "product": Product {
        "name": "toothbrush",
        "unit": 1,
      },
      "quantity": 3,
      "totalPrice": 2.9699999999999998,
    },
  ],
}
`;

exports[`Supermarket applies twenty percent discount 1`] = `
Receipt {
  "discounts": Array [
    Discount {
      "description": "20% off",
      "discountAmount": 0.39799999999999996,
      "product": Product {
        "name": "apples",
        "unit": 0,
      },
    },
  ],
  "items": Array [
    ReceiptItem {
      "price": 1.99,
      "product": Product {
        "name": "apples",
        "unit": 0,
      },
      "quantity": 1,
      "totalPrice": 1.99,
    },
  ],
}
`;

exports[`Supermarket applies two for amount discount 1`] = `
Receipt {
  "discounts": Array [
    Discount {
      "description": "2 for 0.99",
      "discountAmount": 0.3899999999999999,
      "product": Product {
        "name": "cherry tomatoes",
        "unit": 1,
      },
    },
  ],
  "items": Array [
    ReceiptItem {
      "price": 0.69,
      "product": Product {
        "name": "cherry tomatoes",
        "unit": 1,
      },
      "quantity": 2,
      "totalPrice": 1.38,
    },
  ],
}
`;
